hack/tools/conversion-verifier/main.go

```diff
@@ -0,0 +1,196 @@
+
+	apiPackages         = map[string]map[string]*loader.Package{}
+	storageVersionTypes = map[string]*storageVersionType{}
+)
+
+type versionType struct {
+	base string
+	pkg  *loader.Package
+	typ  *markers.TypeInfo
+}
+
+func (v *versionType) IsConvertible() bool {
+	return hasMethod(v.pkg, v.typ, "ConvertFrom") && hasMethod(v.pkg, v.typ, "ConvertTo")
+}
+
+type storageVersionType struct {
+	*versionType
+
+	otherDecls map[string]*versionType
+}
+
+func (s *storageVersionType) IsHub() bool {
+	return hasMethod(s.pkg, s.typ, "Hub")
+}
+
+func main() {
+	var result error
+
+	// Define the marker collector.
+	col := &markers.Collector{
+		Registry: &markers.Registry{},
+	}
+	// Register the markers.
+	if err := col.Registry.Register(storageVersion); err != nil {
+		klog.Fatal(err)
+	}
+	if err := col.Registry.Register(nolintConversion); err != nil {
+		klog.Fatal(err)
+	}
+
+	// Load all packages.
+	packages, err := loader.LoadRoots("./...")
+	if err != nil {
+		klog.Fatal(err)
+	}
+
+	// First loop through all the packages
+	// and find the ones that have a storage version assigned.
+	for i := range packages {
+		root := packages[i]
+
+		base := basePackage(root.PkgPath)
+		if !strings.HasSuffix(base, "/api") {
+			// We're only interested in api folders, exclude everything
+			// else that's not an API package.
+			continue
+		}
+
+		// Store the packages.
+		if _, ok := apiPackages[base]; !ok {
+			apiPackages[base] = make(map[string]*loader.Package)
+		}
+		pkgs := apiPackages[base]
+		pkgs[root.PkgPath] = root
+
+		// Populate type information and loop through every type.
+		root.NeedTypesInfo()
+		if err := markers.EachType(col, root, func(info *markers.TypeInfo) {
+			// Check if the type is registered with storage version.
+			if m := info.Markers.Get(storageVersion.Name); m == nil {
+				return
+			}
+			if _, ok := storageVersionTypes[info.Name]; ok {
+				result = multierror.Append(result,
+					errors.Errorf("type %q has a redeclared storage version in package %+v", info.Name, root.PkgPath),
+				)
+				return
+			}
+
+			// Store the type and package as selected storage version.
+			storageVersionTypes[info.Name] = &storageVersionType{
+				versionType: &versionType{
+					base: base,
+					pkg:  root,
+					typ:  info,
+				},
+				otherDecls: make(map[string]*versionType),
+			}
+		}); err != nil {
+			klog.Fatal(err)
+		}
+	}
+
+	// Now that we have storage version information,
+	// loop through every package again and collect info about all declarations.
+	for _, childs := range apiPackages {
+		for i := range childs {
+			root := childs[i]
+
+			if err := markers.EachType(col, root, func(info *markers.TypeInfo) {
+				ex, ok := storageVersionTypes[info.Name]
+				if !ok {
+					return
+				}
+				if root.PkgPath == ex.pkg.PkgPath {
+					return
+				}
+				// Check if the type is registered with storage version.
```

Comment:
Is this comment correct? Isn't it like check if the nolint marker is set?

Reply:
No, copy paste